[NEW] blog post (#3246)
huggingface/blog: open-responses.md

---
title: "Open Responses: What we need to know" 
thumbnail: /blog/assets/openresponses/thumbnail.png
authors:
- user: evalstate  
- user: burtenshaw  
- user: merve
- user: pcuenq
---

<!-- TODO: @burtenshaw: fancy thumbnail -->

> Open Responses is a new and open inference standard. Initiated by OpenAI, built by the open source AI community, and backed by the Hugging Face ecosystem. Open Responses is based on the Responses API and fits for the future of Agents. In this blog post, we’ll look at how Open Responses works and why the open source community should use Open Responses.

Reviewer: ```suggestion
Open Responses is a new and open inference standard for LLM requests. Initiated by OpenAI, built by the open source AI community, and backed by the Hugging Face ecosystem, Open Responses is based on the Responses API and is designed for the future of Agents. In this blog post, we’ll look at how Open Responses works and why the open source community should use Open Responses.
```

The `for LLM requests` is obvious perhaps, feel free to ignore.

> The era of the chatbot is long gone, and agents dominate inference workloads. Developers are shifting toward autonomous systems that reason, plan, and act over long time horizons. Despite this shift, much of the ecosystem still uses the **Chat Completion** format, which is designed for turn-based conversations and falls short for agentic use cases compared to the **Responses format**, which is closed and not as widely adopted. The **Chat Completion** format has become a de facto standard despite the alternatives.

Reviewer 2: After reading the blog post and the examples, I understand _how_ Open Responses differs from Chat Completions, but I didn't get a clear sense of the advantages of it for agentic workflows. Perhaps an example that highlights something which was clunky in Chat Completions but clean in Open Responses would help get the point across?

- Create Video content through a separate task-based endpoint  
- Run agentic loops on the provider side, executing tool calls autonomously and returning the final result.

# What is Open Responses?

Open Responses extends and open-sources the Responses API,  making it more accessible for builders and routing providers to interoperate and collaborate on shared interests. 

Some of the key points are:

- Stateless by default, supporting encrypted reasoning for providers that require it.  
- Standardized model configuration parameters.  
- Streaming is modeled as a series of semantic events, not raw text or object deltas.  
- Works as both a routing layer and a model provider endpoint. Routers can provide custom tools and agent loops, and forward requests to upstream providers.

## What do we need to know to build with Open Responses?

We’ll briefly explore the core changes that impact most community members. If you want to deep dive into the specification, check out the [Open Responses documentation](https://www.openresponses.org/).

### Client Requests to Open Responses

Client requests to Open Responses are similar to the existing Responses API:

````shell

```diff

>  curl https://evalstate-openresponses.hf.space/v1/responses \

Reviewer 2: Might be good to explain that you've built a proxy or dedicate Space to demo this

   -H "Content-Type: application/json" \
   -H "Authorization: Bearer $HF_TOKEN" \
+  -H "OpenResponses-Version: latest" \
   -N \
   -d '{
         "model": "moonshotai/Kimi-K2-Thinking:nebius",
         "input": "explain the theory of life"
       }'
```
````

### Changes for Inference Clients and Providers

Clients that already support the Responses API can migrate to Open Responses with relatively little effort. The main changes are:

- Migrating reasoning streams to use extendable “reasoning” chunks rather than “reasoning\_text”.  

> - Implementing richer state changes and payloads \- for example, a hosted Code Interpreter can send a specific \`interpreting\` state to improve Agent/User observability.

Author: ```suggestion
- Implementing richer state changes and payloads - for example, a hosted Code Interpreter can send a specific `interpreting` state to improve Agent/User observability.
```

Over time, as Providers continue to innovate, certain features will become standardized in the base specification.

In summary, migrating to Open Responses will make the inference experience more consistent and improve quality as undocumented extensions, interpretations, and workarounds of the legacy Completions API are normalized in Open Responses.

You can see how to stream reasoning chunks below.

```json
 {
  "model": "moonshotai/Kimi-K2-Thinking:together",
  "input": [
    {
      "type": "message",
      "role": "user",
      "content": "explain photosynthesis."
    }
  ],
  "stream": true
}

```

Here’s the difference between Open Response and Responses for reasoning deltas:

```diff
{
  "delta": " heres what i'm thinking",
  "sequence_number": 12,
+ "type": "response.reasoning.delta",
- "type": "response.reasoning_text.delta",
  "item_id": "msg_cbfb8a361f26c0ed0cb133b3c2387279b3d54149a262f3a7",
  "output_index": 0,
  "obfuscation": "0HG8OhAdaLQBg",
  "content_index": 0
}
```

### Open Responses for Routing

Open Responses distinguishes between “Model Providers” \- those who provide inference, and “Routers” \- intermediaries who orchestrate between multiple providers. 

> Clients can now specify a Provider along with provider-specific API options when making requests, allowing intermediary Routers to orchestrate requests between upstream providers.

Reviewer 3: These were removed from spec @pcuenca

![process diagram][https://huggingface.co/huggingface/documentation-images/resolve/main/openresponses/image1.png]

The loop operates as follows:

1. The API receives a user request and samples from the model  
2. If the model emits a tool call, the API executes it (internally or externally)  
3. Tool results are fed back to the model for continued reasoning  
4. The loop repeats until the model signals completion

For internally-hosted tools, the provider manages the entire loop; executing tools, returning results to the model, and streaming output. This means that multi-step workflows like "search documents, summarize findings, then draft an email" use a single request.

Clients control loop behavior via `max_tool_calls` to cap iterations and `tool_choice` to constrain which tools are invocable:

json

```json
{
  "model": "zai-org/GLM-4.7",
  "input": "Find Q3 sales data and email a summary to the team",
  "tools": [...],
  "max_tool_calls": 5,
  "tool_choice": "auto"
}
```

The response contains all intermediate items: tool calls, results, reasoning.

## Next Steps

> Open Responses extends and improves the Responses API, providing richer and more detailed content definitions, compatibility, and deployment options. It also provides a standard way to execute sub-agent loops during primary inference calls, opening up powerful capabilities for AI Applications. We are looking forward to working with the Open Responses team **and community** on future development of the specification. 

Reviewer: ```suggestion
Open Responses extends and improves the Responses API, providing richer and more detailed content definitions, compatibility, and deployment options. It also provides a standard way to execute sub-agent loops during primary inference calls, opening up powerful capabilities for AI Applications. We are looking forward to working with the Open Responses team **and the community at large** on future development of the specification. 
```
